[MOO] add new dynamic optimization to C runtime

linuslangenkamp opened the pull request:
### Purpose

- Add first, experimental version of the new dynamic optimization runtime 
- Interfaced with the new generic library MOO (https://github.com/AMIT-HSBI/MOO), which has been added to 3rdParty

### Discussion
- The library and its C runtime implementation is written in C++ 17. The build can be turned off by setting
`-DOM_OMC_ENABLE_MOO=OFF`. For the PR I have set it to `ON`
- The entry point for the new optimization runtime is located at the top level inside the generated code (see `OMCompiler/Compiler/Template/CodegenC.tpl`), which may be argued about. However, I believe it is better than placing it inside the `solveMain()` routine, as is currently the case

### Note

- Current optimization runtime remains intact even if `-DOM_OMC_ENABLE_MOO=ON`
- The library and the interface are still experimental and a lot of features might not work

### How to run Optimizations with MOO?

- Optimizations can be run by setting the flag `-moo`. In addition grammar must be set to `+g=Optimica`
 -> This is done automatically when calling the command `moo()` inside a .mos file
- MOO introduces a new adaptive mesh refinement algorithm (paper is accepted to the 16th International Modelica & FMI Conference), which can be accessed via `-moo_l2bn_*`
- MOO also has a verification system: Here the optimal controls are used as input to a simulation to verify validity

Simple example:

```modelica
model hypersensitive_moo
  Real x (start=1, fixed=true);
  input Real u;
  output Real finalConstraint(min = 1.5, max=1.5) = x annotation(isFinalConstraint=true);
  output Real lagrangeTerm = 0.5 * (x^2 + u^2) annotation(isLagrange=true);

equation
  der(x) = -x^3 + u;

annotation(
    experiment(StartTime = 0, StopTime = 10000, Tolerance = 1e-12, Interval = 1000), 
     __OpenModelica_commandLineOptions = "+g=Optimica", __OpenModelica_simulationFlags(moo = "()", lv = "LOG_STDOUT,LOG_ASSERT,LOG_STATS", moo_l2bn_p1_it = "3", moo_l2bn_p2_it = "10", moo_l2bn_p2_lvl = "0.0", optimizerNP = "7", variableFilter = ".*", s = "gbode"));
end hypersensitive_moo;
```

### To do

- I have managed to get the interface running with OMEdit, but I still need to implement proper logging such that messages from the optimizer (Ipopt) and MOO can be accessed in real time - for now only in the log file as they are written into stdout
- Write an Optimization module for the NB:
-> Get better structures in the interface
-> More generic optimization problems (e.g. parameter optimizations). The MOO library natively support them, e.g. for training Physics-enhanced Neural ODEs or optimizing systems with free static parameters, but we dont have all parameter derivatives at the moment
-> The optimizations require 2nd order derivative information. I have implemented a new numerical Hessian (via finite differences), but a symbolic Hessian will yield way faster and more stable results

phannebohm commented on OMCompiler/SimulationRuntime/c/simulation/simulation_runtime.cpp:
```diff
@@ -1358,6 +1358,15 @@ int _main_SimulationRuntime(int argc, char**argv, DATA *data, threadData_t *thre
   return retVal;
 }
 
+#ifndef OMC_HAVE_MOO
+
+int _main_OptimizationRuntime(int argc, char**argv, DATA *data, threadData_t *threadData) {
+  errorStreamPrint(OMC_LOG_STDOUT, 0, "MOO has not been built and can not be called: Set -DOM_OMC_ENABLE_MOO=ON to build MOO.");
+  return -1;
+}
+
+#endif // OMC_HAVE_MOO
+
```
Better put this next to the real implementation in OMCompiler/SimulationRuntime/c/moo/main_opt.cpp

linuslangenkamp replied:
I exclude all files in c/moo/*, if moo is not enabled, so i cant do that right now. But maybe I should just create a c/moo/fallback.cpp or so and add it if MOO was not built

linuslangenkamp commented:
I will need to look into the configure builds. For now, these are not available and MOO can only be built with cmake. However, I think we should support configure also.

linuslangenkamp commented:
Somehow the `get.Mumps` (download Mumps sources) script does not terminate correctly and leads to missing sources, when MOO runs its CMake configure step (perhaps because `gunzip` or a similar command is not executing properly).

To work around this, I added the MUMPS linear solver sources to 3rdParty and updated the PR. Hopefully, this allows the MUMPS and Ipopt configure steps to proceed further.

Otherwise, we may need proper CMake builds of MUMPS and Ipopt, which would be painful.

adrpo commented:
Yeah, the Win/UCRT64-qt6 build is not a problem, as we are missing qtwebengine on windows but the cmake on Windows seems to have issues which yeah, seems to be related to MUMPS.

adrpo commented:
I will leave the CI/CMake/Enable/MSYS2-UCRT64 label on, I removed the other one that already works OK.

linuslangenkamp commented:
I have no clue why the Ipopt build does not find the auxiliary autoconf scripts. The MUMPS build has the exact same structure, requires its own (almost identical) autoconf scripts, and finds them correctly. The MUMPS build only fails with LAPACK linker error, which should be an easy fix, but that Ipopt error seems very cryptic.

I will need a proper windows setup and reproduce the error.

adrpo commented:
@linuslangenkamp you can merge this.
